tests/cpp/test_multidevice_matmul.cpp

```diff
@@ -404,4 +404,167 @@ TEST_F(DistributedMatmulTest, Matmul_LayoutNT_ReduceScatter) {
                                     ->heuristic();
   EXPECT_EQ(heuristic, ScheduleHeuristic::ExprEval);
 }
+
+TEST_F(DistributedMatmulTest, MLP_Layer) {
+  std::unique_ptr<Fusion> fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+  auto mesh = DeviceMesh::createForNumDevices(communicator_->size());
+
+  int64_t sb = 64; // sequence * batch
+  int64_t h = 128;
+  int64_t h4 = 4 * h;
+
+  // TODO: error with dynamic shape
+  // C++ exception with description "ext_opt.hasValue() INTERNAL ASSERT FAILED
+  // at "csrc/dynamic_transform.cpp":276, Could not evaluate dynamic extent: i3
+  // Exception raised from DynamicTransformConcretizationInfo at
+  // csrc/dynamic_transform.cpp:276
+  TensorView* x = makeContigConcreteTensor({sb, h}, DataType::BFloat16);
+  TensorView* w0 = makeContigConcreteTensor(
+      {num_devices_, h4 / num_devices_, h}, DataType::BFloat16);
+  TensorView* b0 = makeContigConcreteTensor(
+      {num_devices_, h4 / num_devices_}, DataType::BFloat16);
+  TensorView* w1 = makeContigConcreteTensor(
+      {num_devices_, h, h4 / num_devices_}, DataType::BFloat16);
+  TensorView* b1 = makeContigConcreteTensor({h}, DataType::BFloat16);
+  fusion->addInput(x);
+  fusion->addInput(w0);
+  fusion->addInput(b0);
+  fusion->addInput(w1);
+  fusion->addInput(b1);
+
+  // Linear #1
+  // Notes: Manually breaking down the linear layer into nvfuser primitives
+  TensorView* linear_int0 = broadcast(x, {true, false, true, false});
+  TensorView* linear_int1 = broadcast(w0, {false, true, false, false});
+  TensorView* linear_int2 = mul(linear_int0, linear_int1);
+  TensorView* linear_int3 = sum(linear_int2, {-1});
+  TensorView* linear_int4 = broadcast(b0, {false, true, false});
+  TensorView* linear1 = add(linear_int3, linear_int4);
+
+  TensorView* linear1_ = castOp(DataType::Float, linear1);
+  TensorView* gelu = tanh_gelu(linear1_);
+  TensorView* gelu_ = castOp(DataType::BFloat16, gelu);
+
+  // Linear #2
+  gelu_ = segment_set(gelu_);
```

Comment:
Can you comment on why this is needed?

Reply:
Without the segment set, the reduction scheduler gets called instead of the matmul scheduler. Can add a comment in the code as well.

Reply:
btw, IIUC we should use `matmul` op for both nvfuser matmul and aten matmul in the future and this will address the segmentation issue we see here. We will just need to update our resharding passes to handle matmul and linear ops appropriately.

Reply:
> Can add a comment in the code as well.

That'd always be helpful. Thank you! 

> update our resharding passes to handle matmul and linear ops appropriately

I'm surprised using matmul/linear changes resharding at all because they are done locally. What do you mean?

Reply:
Ahh I should clarify, it would be in the `insertResharding` pass which automatically adds `set` operations where necessary. The work around is very simple,  we just add the `set` manually.

Reply:
SG. I'm fine leaving the segment_set as is. However, in the next (or next next :) PR, we should improve insertResharding to avoid the segment_set. Thunder or nvFuser's Python API doesn't have segment_set and it's too low-level for the framework to add segment_set properly anyway. How does that sound?

Reply:
If we use matmul instead of the broadcast+mul-sum no segment_set is needed, so in the next PR we will have a version with no segment set :)